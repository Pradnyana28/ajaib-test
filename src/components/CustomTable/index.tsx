import React from 'react';
import {
    Table,
    Thead,
    Tbody,
    Tr,
    Th,
    Td,
    TableContainer,
} from '@chakra-ui/react'
import SearchAndFilter, { FILTER_DEFAULT_VALUE } from '../SearchAndFilter/SearchAndFilter';

interface RandomResponse {
    info: {
        page: number;
        results: number;
        seed: string;
        version: string;
    };
    results: RandomData[];
}

export interface RandomData {
    name: {
        title: string;
        first: string;
        last: string;
    };
    email: string;
    gender: string;
    login: {
        uuid: string;
        username: string;
        password: string;
        salt: string;
        sha1: string;
        sha256: string;
        md5: string;
    },
    registered: {
        date: string;
        age: number;
    }
}

interface TableData {
    label: string;
    isSearchable: boolean;
    value?: string;
    isNumeric?: boolean;
};

export interface TableItem {
    [field: string]: TableData;
};

export interface CustomTableProps {
    fetchEndpoint: string;
    dataMapping: (data: RandomData) => Record<string, Record<string, any>>;

    getSearchValue?: React.Dispatch<React.SetStateAction<any>> | ((value: string) => void);
    getFilterValue?: React.Dispatch<React.SetStateAction<any>> | ((value: string) => void);
}

const fetchAPI = async (endpoint: string, options: Record<string, string | undefined>) => {
    const queryString = new URLSearchParams({
        ...options,
        results: String(10),
        page: String(1)
    });
    const response = await fetch(`${endpoint}?${queryString}`);

    if (response.ok) {
        return await response.json() as RandomResponse;
    }
};

const findColumns = (columns: any[], value: string) => columns.map(field => {
    const searchFields = Object.keys(field).filter(f =>
        field[f].isSearchable
        && field[f].value
            .toString()
            .toLowerCase()
            .includes(value.toLowerCase())
    );
    return searchFields.length ? field : undefined;
}).filter(x => x);

const TableContent = ({ data }: { data: TableItem[] }) => {
    const firstItem = data[0];

    const _onHandleSort = (field: string) => {
        console.log(field);
    }

    return (
        <>
            <Thead>
                <Tr>
                    {Object.keys((firstItem)).map(d => <Th key={d} isNumeric={data[0][d].isNumeric} onClick={() => _onHandleSort(d)}>{d}</Th>)}
                </Tr>
            </Thead>
            <Tbody>
                {data.map((d, idx) => (
                    <Tr key={idx}>
                        {Object.keys(d).map(item => <Td key={item} isNumeric={d[item].isNumeric}>{d[item].value}</Td>)}
                    </Tr>
                ))}
            </Tbody>
        </>
    )
}

const TableWrapper = ({ columns }: { columns: any[] }) => {
    return columns?.length ? (
        <TableContainer>
            <Table variant='simple'>
                <TableContent data={columns} />
            </Table>
        </TableContainer>
    ) : <p>No data found.</p>;
}

const CustomTable = (props: CustomTableProps) => {
    const [columns, setColumns] = React.useState<any[]>([]);
    const [filteredColumns, setFilteredColumns] = React.useState<any[]>([]);
    const [searchValue, setSearchValue] = React.useState(null);
    const [filterValue, setFilterValue] = React.useState(FILTER_DEFAULT_VALUE);

    React.useEffect(() => {
        if (filterValue === 'all') {
            fetchUsers();
        } else {
            fetchUsers(filterValue);
        }
    }, [filterValue]);

    React.useEffect(() => {
        if (searchValue !== null) {
            setTimeout(() => {
                searchColumns(searchValue);
            }, 750);
        }
    }, [searchValue]);

    const fetchUsers = async (gender?: string) => {
        const data = await fetchAPI(props.fetchEndpoint, { gender });
        if (data) {
            const transformedData = data.results.map(d => props.dataMapping(d));
            setColumns(transformedData);
            setFilteredColumns(transformedData);
        }
    }

    const searchColumns = async (value: string) => {
        if (value === '') {
            setFilteredColumns(columns);
        } else {
            const searchableFields = findColumns(columns, value);
            setFilteredColumns(searchableFields);
        }
    }

    return (
        <>
            <SearchAndFilter getSearchValue={setSearchValue} getFilterValue={setFilterValue} />
            <TableWrapper columns={filteredColumns} />
        </>
    )
}

export default CustomTable;