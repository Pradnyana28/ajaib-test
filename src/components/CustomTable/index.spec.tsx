import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import CustomTable from '.';

global.fetch = jest.fn(() =>
    Promise.resolve({
        json: () => Promise.resolve({ test: 100 }),
    }),
) as jest.Mock;

describe('<CustomTable />', () => {
    it('renders without crashing', async () => {
        const { container } = render(<CustomTable fetchEndpoint='' dataMapping={(data) => ([])} />);
        expect(container).toMatchSnapshot();
    });
});