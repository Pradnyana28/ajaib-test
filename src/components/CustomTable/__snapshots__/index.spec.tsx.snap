// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CustomTable /> renders without crashing 1`] = `
<div>
  <div
    class="chakra-stack css-17kdbol"
  >
    <div
      class="css-147dy5z"
    >
      <div>
        <p
          class="chakra-text css-g5hezf"
        >
          Search
        </p>
        <div
          class="chakra-input__group css-4302v8"
        >
          <input
            class="chakra-input css-tsqr3a"
            data-testid="search-input"
            placeholder="Search Keywords"
            value=""
          />
          <div
            class="chakra-input__right-addon css-nxc9bd"
          >
            <svg
              class="chakra-icon css-onkibi"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M23.384,21.619,16.855,15.09a9.284,9.284,0,1,0-1.768,1.768l6.529,6.529a1.266,1.266,0,0,0,1.768,0A1.251,1.251,0,0,0,23.384,21.619ZM2.75,9.5a6.75,6.75,0,1,1,6.75,6.75A6.758,6.758,0,0,1,2.75,9.5Z"
                fill="currentColor"
              />
            </svg>
          </div>
        </div>
      </div>
      <div>
        <p
          class="chakra-text css-g5hezf"
        >
          Gender
        </p>
        <div
          class="chakra-select__wrapper css-42b2qy"
        >
          <select
            class="chakra-select css-1ik61og"
            data-testid="gender-filter"
          >
            <option
              value=""
            >
              Select Gender
            </option>
            <option
              data-testid="gender-filter-all"
              value="all"
            >
              All
            </option>
            <option
              data-testid="gender-filter-male"
              value="male"
            >
              Male
            </option>
            <option
              data-testid="gender-filter-female"
              value="female"
            >
              Female
            </option>
          </select>
          <div
            class="chakra-select__icon-wrapper css-162mkon"
          >
            <svg
              aria-hidden="true"
              class="chakra-select__icon"
              focusable="false"
              role="presentation"
              style="width: 1em; height: 1em; color: currentColor;"
              viewBox="0 0 24 24"
            >
              <path
                d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
                fill="currentColor"
              />
            </svg>
          </div>
        </div>
      </div>
      <div>
        <p
          class="chakra-text css-g5hezf"
        >
          Â 
        </p>
        <button
          class="chakra-button css-4xx2wk"
          data-testid="reset-filter-button"
          type="button"
        >
          Reset Filter
        </button>
      </div>
    </div>
    <p>
      No data found.
    </p>
    <div
      class="css-k634du"
    >
      <p>
        Page 
        <b>
          1
        </b>
         of 
        513
      </p>
      <div
        class="css-17xejub"
      />
      <div
        class="chakra-stack css-193rmy8"
      >
        <button
          aria-label="Previous page"
          class="chakra-button css-p7mkev"
          data-active=""
          type="button"
        >
          <svg
            aria-hidden="true"
            class="chakra-icon css-onkibi"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M15.41 7.41L14 6l-6 6 6 6 1.41-1.41L10.83 12z"
              fill="currentColor"
            />
          </svg>
        </button>
        <button
          class="chakra-button css-4xx2wk"
          data-active=""
          type="button"
        >
          1
        </button>
        <button
          class="chakra-button css-4xx2wk"
          type="button"
        >
          2
        </button>
        <button
          class="chakra-button css-4xx2wk"
          type="button"
        >
          3
        </button>
        <button
          class="chakra-button css-4xx2wk"
          type="button"
        >
          4
        </button>
        <button
          class="chakra-button css-4xx2wk"
          type="button"
        >
          5
        </button>
        <button
          class="chakra-button css-4xx2wk"
          type="button"
        >
          6
        </button>
        <p>
          ...
        </p>
        <button
          class="chakra-button css-4xx2wk"
          type="button"
        >
          509
        </button>
        <button
          class="chakra-button css-4xx2wk"
          type="button"
        >
          510
        </button>
        <button
          class="chakra-button css-4xx2wk"
          type="button"
        >
          511
        </button>
        <button
          class="chakra-button css-4xx2wk"
          type="button"
        >
          512
        </button>
        <button
          class="chakra-button css-4xx2wk"
          type="button"
        >
          513
        </button>
        <button
          aria-label="Next page"
          class="chakra-button css-p7mkev"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="chakra-icon css-onkibi"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M10 6L8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6z"
              fill="currentColor"
            />
          </svg>
        </button>
      </div>
    </div>
  </div>
</div>
`;
